worlds/sa2b/Regions.py

```diff
@@ -511,35 +596,101 @@ def connect_regions(world, player, gates: typing.List[LevelGate], cannon_core_em
+                    lambda state: (state.has(ItemName.knuckles_shovel_claws, player)))
+        else:
+            connect(world, player, names, LocationName.gate_2_boss_region, LocationName.gate_2_region)
+
         for i in range(len(gates[2].gate_levels)):
             connect(world, player, names, LocationName.gate_2_region, shuffleable_regions[gates[2].gate_levels[i]])
 
     if len(gates) >= 4:
-        connect(world, player, names, 'Menu', LocationName.gate_3_region,
+        connect(world, player, names, 'Menu', LocationName.gate_3_boss_region,
                 lambda state: (state.has(ItemName.emblem, player, gates[3].gate_emblem_count)))
+
+        if gate_bosses[3] == all_gate_bosses_table[king_boom_boo]:
+            connect(world, player, names, LocationName.gate_3_boss_region, LocationName.gate_3_region,
+                    lambda state: (state.has(ItemName.knuckles_shovel_claws, player)))
+        else:
+            connect(world, player, names, LocationName.gate_3_boss_region, LocationName.gate_3_region)
+
         for i in range(len(gates[3].gate_levels)):
             connect(world, player, names, LocationName.gate_3_region, shuffleable_regions[gates[3].gate_levels[i]])
 
     if len(gates) >= 5:
-        connect(world, player, names, 'Menu', LocationName.gate_4_region,
+        connect(world, player, names, 'Menu', LocationName.gate_4_boss_region,
                 lambda state: (state.has(ItemName.emblem, player, gates[4].gate_emblem_count)))
+
+        if gate_bosses[4] == all_gate_bosses_table[king_boom_boo]:
+            connect(world, player, names, LocationName.gate_4_boss_region, LocationName.gate_4_region,
+                    lambda state: (state.has(ItemName.knuckles_shovel_claws, player)))
+        else:
+            connect(world, player, names, LocationName.gate_4_boss_region, LocationName.gate_4_region)
+
         for i in range(len(gates[4].gate_levels)):
             connect(world, player, names, LocationName.gate_4_region, shuffleable_regions[gates[4].gate_levels[i]])
 
     if len(gates) >= 6:
-        connect(world, player, names, 'Menu', LocationName.gate_5_region,
+        connect(world, player, names, 'Menu', LocationName.gate_5_boss_region,
                 lambda state: (state.has(ItemName.emblem, player, gates[5].gate_emblem_count)))
+
+        if gate_bosses[5] == all_gate_bosses_table[king_boom_boo]:
+            connect(world, player, names, LocationName.gate_5_boss_region, LocationName.gate_5_region,
+                    lambda state: (state.has(ItemName.knuckles_shovel_claws, player)))
+        else:
+            connect(world, player, names, LocationName.gate_5_boss_region, LocationName.gate_5_region)
+
         for i in range(len(gates[5].gate_levels)):
             connect(world, player, names, LocationName.gate_5_region, shuffleable_regions[gates[5].gate_levels[i]])
 
+    if len(gates) == 1:
```

Comment:
this should be an elif tree or a dict lookup

Reply:
Changed this to an elif tree.